cli/azd/pkg/infra/provisioning/bicep/bicep_cache.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package bicep
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/azure/azure-dev/cli/azd/pkg/azure"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment"
+	"github.com/azure/azure-dev/cli/azd/pkg/environment/azdcontext"
+	"github.com/azure/azure-dev/cli/azd/pkg/lazy"
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"golang.org/x/exp/slices"
+)
+
+// cache is serialized into json and represents the local provisioning cache for bicep.
+type cache struct {
+	Template   json.RawMessage
+	Parameters azure.ArmParameters
+}
+
+// CacheManager provides the functionality for using bicep cache.
+type CacheManager interface {
+	// Current returns existing cache or nil when there's no cache.
+	Current(context context.Context) *cache
+	// Cache persist cache. Use Cache(nil) to clear cache.
+	Cache(context context.Context, cache *cache) error
+	// Equal compares cache against current and return false when they are different.
+	Equal(context context.Context, cache *cache) bool
+}
```

Comment:
I would expect this `CacheManager` to use public components and right now `cache` is private.  Let's think on how we could make this cache manager live at the root of azd and not be tied directly to provisioning.  At that point I would expect to see a generic cache manager where maybe we take a keyed based approach to get/set cache items with a particular key & value that then get persisted to the cache stores.